mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -172,37 +255,63 @@ object StringIndexer extends DefaultParamsReadable[StringIndexer] {
 
   @Since("1.6.0")
   override def load(path: String): StringIndexer = super.load(path)
+
+  // Returns a function used to sort strings by frequency (ascending or descending).
+  // In case of equal frequency, it sorts strings by alphabet (ascending).
+  private[feature] def getSortFunc(
+      ascending: Boolean): ((String, Long), (String, Long)) => Boolean = {
+    (a: (String, Long), b: (String, Long)) => {
+      if (a._2 == b._2) {
+        a._1 < b._1
+      } else {
+        if (ascending) a._2 < b._2 else a._2 > b._2
```

Comment:
This is minor and a nit given it's use here, and the branch prediction will take care of this quickly -- but we could move the check and return two different functions (or just define two different sort functions) and avoid having the extra branch during the sort). But in other places it would be more important to optimize this so just a side comment while I'm reading through.